src/main/java/fi/aalto/cs/intellij/services/CourseInformation.java

```diff
@@ -0,0 +1,106 @@
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CourseInformation {
+  private static final Logger logger = LoggerFactory
+      .getLogger(CourseInformation.class);
+
+  private String courseName;
+  private Map<String, URL> moduleUrls;
+
+  public CourseInformation() {
+    // Replace this with the correct path when testing with a local course configuration file.
+    parse("o1.json");
+  }
+
+  /**
+   * The CourseInformation instance managed by the service manager.
+   */
+  public static CourseInformation getInstance() {
+    return ServiceManager.getService(CourseInformation.class);
+  }
+
+  /**
+   * Parses the given course configuration file.
+   * @param pathToCourseConfig The path to the course configuration file.
+   */
+  public void parse(String pathToCourseConfig) {
+    logger.info("Parsing the course configuration file '{}'", pathToCourseConfig);
+    courseName = "";
+    moduleUrls = new HashMap<>();
+    FileReader file;
+    try {
+      file = new FileReader(pathToCourseConfig);
+    } catch (FileNotFoundException e) {
+      // TODO: notify the user of the error with a notification
+      logger.error("Failed to find course configuration file '{}'", pathToCourseConfig);
+      return;
+    }
+
+    JSONTokener tokenizer = new JSONTokener(file);
+    JSONObject jsonObj = new JSONObject(tokenizer);
+    try {
+      courseName = jsonObj.getString("name");
+
+      JSONObject modules = jsonObj.getJSONObject("modules");
+      Iterator<String> it = modules.keys();
+      while (it.hasNext()) {
+        String key = it.next();
+        String urlString = modules.getJSONObject(key).getString("url");
+        URL moduleUrl;
+        try {
+          moduleUrl = new URL(urlString);
+        } catch (MalformedURLException e) {
+          // TODO: notify the user of the error with a notification
+          logger.error("Invalid url '{}' in course configuration file", urlString);
+          return;
+        }
+        moduleUrls.put(key, moduleUrl);
+      }
+    } catch (JSONException e) {
+      // TODO: notify the user of the error with a notification
+      logger.error("Invalid JSON in course configuration file: {}", e.getMessage());
+    }
+  }
+
+  /**
+   * Returns the name of the currently loaded course.
+   */
+  public String getCourseName() {
+    return courseName;
+  }
+
+  /**
+   * Returns the names of the code modules for the currently loaded course.
+   */
+  public List<String> getModuleNames() {
+    List<String> moduleNames = new ArrayList<String>();
+    moduleNames.addAll(moduleUrls.keySet());
```

Comment:
Good catch! It's certainly desireable to list the modules in the order that they are in the course config file, so I'll make the order deterministic here and add a mention of the order to the javadoc.

Reply:
Wait... if they are written, in JSON, as an object (or is it called dictionary or map or what but I mean this thing `{...}`) as key-value pairs, then the order does not matter in the JSON either. AFAIK, JSON strings `{"a": 1, "b": 2}` and `{"b": 2, "a": 1}` should be always considered equal, so making the order matter violates this. If the order should matter, either we can store modules in a list/array (`[...]` thing) or use alphabetical/lexicographical order.

Reply:
It would probably work in practice, but yes, if we want to use a specific order, it would be better to use an array and have the names of the modules be part of the module objects stored in the modules array